src/container/mod.rs

```diff
@@ -0,0 +1,226 @@
+            Layout::Horizontal => Self::new_horizontal(children),
+            Layout::Vertical => Self::new_vertical(children),
+            Layout::Grid => Self::new_grid(children),
+        }
+    }
+
+    pub fn new_linear(dir: LinearDir, children: Vec<TileId>) -> Self {
+        Self::Linear(Linear::new(dir, children))
+    }
+
+    pub fn new_horizontal(children: Vec<TileId>) -> Self {
+        Self::new_linear(LinearDir::Horizontal, children)
+    }
+
+    pub fn new_vertical(children: Vec<TileId>) -> Self {
+        Self::new_linear(LinearDir::Vertical, children)
+    }
+
+    pub fn new_tabs(children: Vec<TileId>) -> Self {
+        Self::Tabs(Tabs::new(children))
+    }
+
+    pub fn new_grid(children: Vec<TileId>) -> Self {
+        Self::Grid(Grid::new(children))
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.children().is_empty()
+    }
+
+    pub fn children(&self) -> &[TileId] {
+        match self {
+            Self::Tabs(tabs) => &tabs.children,
+            Self::Linear(linear) => &linear.children,
+            Self::Grid(grid) => &grid.children,
+        }
+    }
+
+    pub fn add_child(&mut self, child: TileId) {
+        match self {
+            Self::Tabs(tabs) => tabs.add_child(child),
+            Self::Linear(linear) => linear.add_child(child),
+            Self::Grid(grid) => grid.add_child(child),
+        }
+    }
+
+    pub fn layout(&self) -> Layout {
+        match self {
+            Self::Tabs(_) => Layout::Tabs,
+            Self::Linear(linear) => match linear.dir {
+                LinearDir::Horizontal => Layout::Horizontal,
+                LinearDir::Vertical => Layout::Vertical,
+            },
+            Self::Grid(_) => Layout::Grid,
+        }
+    }
+
+    pub fn set_layout(&mut self, layout: Layout) {
+        if layout == self.layout() {
+            return;
+        }
+
+        *self = match layout {
+            Layout::Tabs => Self::Tabs(Tabs::new(self.children().to_vec())),
+            Layout::Horizontal => {
+                Self::Linear(Linear::new(LinearDir::Horizontal, self.children().to_vec()))
+            }
+            Layout::Vertical => {
+                Self::Linear(Linear::new(LinearDir::Vertical, self.children().to_vec()))
+            }
+            Layout::Grid => Self::Grid(Grid::new(self.children().to_vec())),
+        };
+    }
+
+    pub(super) fn retain(&mut self, mut retain: impl FnMut(TileId) -> bool) {
+        let retain = |tile_id: &TileId| retain(*tile_id);
+        match self {
+            Self::Tabs(tabs) => tabs.children.retain(retain),
+            Self::Linear(linear) => linear.children.retain(retain),
+            Self::Grid(grid) => grid.children.retain(retain),
+        }
+    }
+
+    pub(super) fn simplify_children(&mut self, mut simplify: impl FnMut(TileId) -> SimplifyAction) {
+        match self {
+            Self::Tabs(tabs) => tabs.children.retain_mut(|child| match simplify(*child) {
```

Comment:
would imho be cleaner if every backing struct of the enum values had a `simplify_children` method